src/rust/lib/callback/src/lib.rs

```diff
@@ -180,6 +197,53 @@ impl SharedRegistryMut {
 }
 
 
+// ==========================
+// === SharedRegistryMut1 ===
+// ==========================
+
+/// Registry gathering callbacks. Each registered callback is assigned with a handle. Callback and
+/// handle lifetimes are strictly connected. As soon a handle is dropped, the callback is removed
+/// as well.
+#[derive(CloneRef,Derivative)]
+#[derivative(Clone(bound=""))]
+#[derivative(Debug(bound=""))]
+#[derivative(Default(bound=""))]
+#[allow(clippy::type_complexity)]
+pub struct SharedRegistryMut1<T> {
+    #[derivative(Debug="ignore")]
+    callback_list: Rc<RefCell<Vec<(Guard,Rc<RefCell<dyn CallbackMut1Fn<T>>>)>>>
+}
+
+impl<T> SharedRegistryMut1<T> {
+    /// Adds new callback and returns a new handle for it.
+    pub fn add<F:CallbackMut1Fn<T>>(&self, callback:F) -> Handle {
+        let callback = Rc::new(RefCell::new(callback));
+        let handle   = Handle::new();
+        let guard    = handle.guard();
+        self.callback_list.borrow_mut().push((guard,callback));
+        handle
+    }
+
+    ///Checks whether there are any callbacks registered.
+    pub fn is_empty(&self) -> bool {
+        self.callback_list.borrow().is_empty()
+    }
+
+    /// Fires all registered callbacks and removes the ones which got dropped. The implementation is
+    /// safe. You are allowed to change the registry while a callback is running.
```

Comment:
If a callback is dropped while some other callback is running, will it be removed?

Reply:
Also, remove comment that `The implementation is safe.` — in itself it says nothing (what exactly is understood by "safe" — my first thought was about Rust safety) and almost suggests that our code is implicitly unsafe unless stated otherwise.

Reply:
The next sentence wa meant as explanation. I will remove dot and use `-` there, ok ?

Reply:
I'm fine with what you suggest, though I'd just remove that safety sentence.
Just say the thing, rather than introduce custom safety notion and then immediately define.